Merge with WebAssembly/gc

dhil opened the pull request:
There is an issue around type substitution to be resolved: tag and continuation types can be either a `VarHT` or a `DefHT`. Previously, `check_heap_type` would make sure there were no `VarHT` during validation, but that's not the case with the infrastructure of the GC proposal. So my question is, what's the best design here, should we return substituted types as in the function references proposal, revert to only indexed types, or something else? I suppose we cannot fully substitute types anymore because of the introduction of recursive types, we can only unroll them a finite amount of times.

rossberg commented:
> Previously, check_heap_type would make sure there were no VarHT during validation, but that's not the case with the infrastructure of the GC proposal.

Hm, validation only deals with "syntactic" types, hence check_heap_type should only encounter types that _only_ contain VarHT, never DefHT. I don't think that has changed.

Recursive types are not an issue, since they are iso-recursive, so they can be substituted just fine without any unrolling.

Unless I'm missing something, I don't think much needs changing for this merge. The only thing I can think of is that places looking up a continuation type in the context now need to invoke expand, like it's done for function and GC types. That's where recursive types get unrolled. Is there another problem you encountered?

dhil commented:
I am seeing non-syntactic continuation types, e.g.
```ocaml
| ContNew x ->
    let ct = cont_type c x in
    let ft = as_func_cont_type ct in
    [RefT (Null, as_heap_str_type (DefFuncT ft))] -->
    [RefT (NoNull, DefHT (type_ c x))], []
```

The `ct` is not a syntactic type, so `as_func_cont_type` triggers an assertion error. I suppose I must be missing some elaboration when the context is being built.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -117,7 +124,8 @@ let check_heap_type (c : context) (t : heap_type) at =
   | VarHT (StatX x) -> let _dt = type_ c (x @@ at) in ()
-  | VarHT (RecX _) | DefHT _ -> assert false
+  | VarHT (RecX _) -> assert false
+  | DefHT _ -> ()
```
I suggest leaving that as assertion failure, it should never happen.

dhil replied:
It was triggered by the exception handling testsuite.. so yet another bug to find!

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -96,68 +113,121 @@ let check_limits {min; max} range at msg =
 let check_tag_type (c : context) (et : tag_type) at =
   match et with
-  | TagT ft -> TagT (check_heap_type c ft at)
+  | TagT ft -> check_heap_type c ft at
```
Nit:
```suggestion
  | TagT ht -> check_heap_type c ht at
```

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -67,19 +67,36 @@ let init_locals (c : context) xs =
+  | DefArrayT at -> at
+  | _ -> error x.at ("non-array type " ^ I32.to_string_u x.it)
+
+let tag_type (c : context) x =
```
This is at odds with the other `xyz_type` helpers, since `x` is not a typeidx. I don't think it should even be needed, the `tag` function should be enough?

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -392,9 +487,9 @@ let rec check_instr (c : context) (e : instr) (s : infer_result_type) : infer_in
-    let tag = tag c x in
-    let tagtype = check_tag_type c tag e.at in
-    let FuncT (ts1, _) = as_func_tag_type tagtype in
+    let tag = tag_type c x in
```
This should just call `tag`, once you have changed as_func_tag_type to take a context, as suggested in the other comment.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -392,9 +487,9 @@ let rec check_instr (c : context) (e : instr) (s : infer_result_type) : infer_in
-    let tagtype = check_tag_type c tag e.at in
-    let FuncT (ts1, _) = as_func_tag_type tagtype in
+    let tag = tag_type c x in
+    check_tag_type c tag e.at;
```
The tag's type has already been checked, no need to check it again.

rossberg commented on interpreter/syntax/types.ml:
```diff
@@ -61,28 +99,198 @@ let is_ref_type = function
+  | DefHT dt -> as_func_str_type (expand_def_type dt)
   | _ -> assert false
 
 let as_func_cont_type (ContT ct) : func_type =
```
These functions aren't really projections, they should perform a lookup in the context to handle Vars. And because they need the context, they need to be defined as helpers in valid.ml, not here. I'd call them func_type_of_heap_type etc, to differentiate them from simple projections.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -508,29 +636,29 @@ let rec check_instr (c : context) (e : instr) (s : infer_result_type) : infer_in
 
   | Suspend x ->
-    let tag = tag c x in
+    let tag = tag_type c x in
```
Same comments as above.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -508,29 +636,29 @@ let rec check_instr (c : context) (e : instr) (s : infer_result_type) : infer_in
     let ct = cont_type c x in
     let FuncT (ts1, ts2) = as_func_cont_type ct in
-    let tag = tag c y in
+    let tag = tag_type c y in
```
And here.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -329,15 +404,35 @@ let check_memop (c : context) (memop : ('t, 's) memop) ty_size get_sz at =
+        | DefContT (ContT ht) -> ht
+        | _ -> assert false)
+    | _ -> assert false
+  in
```
Both these should be unnecessary, if you provide a context argument to the conversion functions below.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -329,15 +404,35 @@ let check_memop (c : context) (memop : ('t, 's) memop) ty_size get_sz at =
     let (_, ts') = label c x2 in
     match Lib.List.last_opt ts'  with
-    | Some (RefT (nul', DefHT (DefContT (ContT ht')))) ->
+    | Some (RefT (nul', ht)) when is_heap_cont_type ht ->
```
You shouldn't need this side condition – it has to be a cont_type. So something like:
```
    | Some (RefT (nul', ht)) ->
      let ht' = cont_type_of_heap_type c ht in
      let ft = func_type_of_cont_type c ht in
```
where the conversion functions produce the error.

(Ideally, I'd like to simplify this whole match expression into a call to `check_stack`, analogous to `BrTable`. Something like:
```
    let FuncT (ts3, ts4) = func_type_of_tag_type c (tag c x1) in
    let ht = DefHT (...) in
    check_stack c (ts3 @ [RefT (NoNull, ht)]) (label c x) x.at
```
But the "..." will get pretty gnarly, and I'm not sure it'd actually work with declared subtyping.)

rossberg reviewed (commented):
I believe the issue is that some of the auxiliary "projections" on tag and cont types assume closed types when they shouldn't. See comments below.

dhil commented:
@rossberg I have implemented your suggested fixes. I am now able to pass the test suite locally. One thing to pay attention to is the new top `ContHT` for continuation types. I am not sure whether we want something else here -- this seemed like the most conservative thing to do to me.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -74,7 +74,7 @@ let func_type (c : context) x =
   match expand_def_type (type_ c x) with
   | DefContT ct -> ct
-  | _ -> error x.at ("non-continuation type " ^ Int32.to_string x.it)
+  | _ as t -> Printf.printf "%s\n%!" (string_of_str_type t); error x.at ("non-continuation type " ^ Int32.to_string x.it)
```
Debug leftover?

dhil replied:
yes!

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -122,10 +137,11 @@ let check_heap_type (c : context) (t : heap_type) at =
   | VarHT (StatX x) -> let _dt = type_ c (x @@ at) in ()
   | VarHT (RecX _) -> assert false
-  | DefHT _ -> ()
+  | DefHT _ -> assert false
```
Nit: combine cases

rossberg commented on test/core/cont.wast:
```diff
@@ -418,129 +418,129 @@
-  (i32.const 4) (i32.const 5) (i32.const 6)
-)
+;; (module $concurrent-generator
+;;   (func $log (import "spectest" "print_i64") (param i64))
```
Accidentally still commented out?

dhil replied:
yes!

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -86,20 +86,35 @@ let array_type (c : context) x =
 
 let refer_func (c : context) x = refer "function" c.refs.Free.funcs x
 
+(* Conversions *)
```
Nit:
```suggestion

(* Conversions *)
```

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -86,20 +86,35 @@ let array_type (c : context) x =
+  func_type_of_heap_type c ht at
+
+let heap_type_of_str_type (_c : context) (st : str_type) : heap_type =
+  DefHT (DefT (RecT [SubT (Final, [], st)], Int32.of_int 0))
```
Nit:
```suggestion
  DefHT (DefT (RecT [SubT (Final, [], st)], Int32.of_int 0))

```

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -86,20 +86,35 @@ let array_type (c : context) x =
+let cont_type_of_heap_type (c : context) (ht : heap_type) at : cont_type =
+  match ht with
+  | DefHT dt -> as_cont_str_type (expand_def_type dt)
+  | VarHT (RecX x | StatX x) -> cont_type c (x @@ at)
```
RecX won't work here, since they are not recorded in the context. They should never occur during checking either, unless you forgot some expand somewhere, so this case should be subsumed by the assert false below.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -404,32 +420,13 @@ let check_memop (c : context) (memop : ('t, 's) memop) ty_size get_sz at =
-      let ht' = extract_cont_heap_type ht in
-      let ft' = as_func_heap_type ht' in
+    | Some (RefT (nul', ht)) ->
+      let ct = cont_type_of_heap_type c ht x2.at in
```
Won't this run into an assertion failure if the program is ill-typed, e.g., the ref type isn't a continuation? (Would be good to have tests for that.)

dhil replied:
I'll write a test for this case.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -86,20 +86,35 @@ let array_type (c : context) x =
+let func_type_of_heap_type (c : context) (ht : heap_type) at : func_type =
+  match ht with
+  | DefHT dt -> as_func_str_type (expand_def_type dt)
+  | VarHT (RecX x | StatX x) -> func_type c (x @@ at)
```
Same here.

rossberg commented on interpreter/exec/eval.ml:
```diff
@@ -177,6 +177,18 @@ let drop n (vs : 'a stack) at =
 
+let str_type_of_heap_type (inst : module_inst) ht : str_type =
+  match ht with
+  | VarHT (StatX x | RecX x) -> str_type inst (x @@ Source.no_region)
```
Same comment re RecX here as in validator: a RecX can't be looked up in the instance; it should never occur if everything is expanded correctly.

dhil commented:
@rossberg I've added a few more tests now. Happy to add more.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -404,32 +420,13 @@ let check_memop (c : context) (memop : ('t, 's) memop) ty_size get_sz at =
-    let FuncT (ts3, ts4) = as_func_tag_type (tag_type c x1) in
+    let FuncT (ts3, ts4) = func_type_of_tag_type c (tag c x1) x1.at in
     let (_, ts') = label c x2 in
     match Lib.List.last_opt ts'  with
```
Nit:
```suggestion
    match Lib.List.last_opt ts' with
```

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -67,23 +67,57 @@ let init_locals (c : context) xs =
+
+let cont_type_of_heap_type (c : context) (ht : heap_type) at : cont_type =
+  match ht with
+  | DefHT dt -> as_cont_str_type (expand_def_type dt)
```
Did you try if this case actually can arise during validation? Because if it can occur, then I'm convinced we can construct a negative test where this runs into an assertion failure.

Can you try removing the case and see if anything breaks?

dhil replied:
Nothing breaks. I think this is just me not having the right model in place, when I wrote this code.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -67,23 +67,57 @@ let init_locals (c : context) xs =
+
+let func_type_of_heap_type (c : context) (ht : heap_type) at : func_type =
+  match ht with
+  | DefHT dt -> as_func_str_type (expand_def_type dt)
```
Similar question here.

dhil replied:
Same as above.

rossberg reviewed (approved):
Looks good to me, except for the question about below.

Would you mind adding the cont/nocont thing, here or in a follow-up, so that it doesn't get lost?

dhil commented:
@rossberg I've added `NoContHT` as a bottom heap type for continuation types (in addition to `ContHT` which I had already added, it is a top type now).

dhil commented:
I've added text support for the various `cont` heap type short hands.

rossberg commented on interpreter/binary/decode.ml:
```diff
@@ -190,6 +190,7 @@ let heap_type s =
       | -0x15 -> StructHT
       | -0x16 -> ArrayHT
       | -0x17 -> ContHT
+      | -0x18 -> NoContHT
```
Can we order this next to the other bottom types, i.e., -0x0c?

dhil replied:
Done. I changed the opcode for `ContHT` to `-0x18` as the other code appears to have been snatched by the exception handling proposal already.

rossberg replied:
Ah, FWIW, that will need a bottom type as well when combined with the GC proposal, so perhaps better even use -0x0b for nocont?

dhil replied:
good idea

rossberg commented on interpreter/valid/match.ml:
```diff
@@ -34,7 +34,7 @@ and bot_of_heap_type c = function
   | FuncHT | NoFuncHT -> NoFuncHT
   | ExternHT | NoExternHT -> NoExternHT
-  | ContHT -> ContHT
+  | ContHT | NoContHT -> NoContHT
```
Right, because this actually was incorrect without the bottom type.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -98,13 +98,13 @@ let refer_func (c : context) x = refer "function" c.refs.Free.funcs x
 let cont_type_of_heap_type (c : context) (ht : heap_type) at : cont_type =
   match ht with
-  | DefHT dt -> as_cont_str_type (expand_def_type dt)
+  | DefHT dt -> assert false
```
Nit: just remove this case, it's subsumed by the default below.

dhil replied:
Thanks! I accidentally forgot to remove this redundant case.

rossberg commented on interpreter/valid/valid.ml:
```diff
@@ -98,13 +98,13 @@ let refer_func (c : context) x = refer "function" c.refs.Free.funcs x
 let func_type_of_heap_type (c : context) (ht : heap_type) at : func_type =
   match ht with
-  | DefHT dt -> as_func_str_type (expand_def_type dt)
+  | DefHT dt -> assert false
```
Same here.

rossberg reviewed (approved):
Great, thanks! Just a couple of nits.
